File: 0200-number-of-islands/0200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    let components = 0
    const rows = grid.length
    const cols = grid[0].length
    for(let r = 0; r < rows; r++){
        for(let c = 0; c < cols; c++){
            if(grid[r][c] == 1){
                helper(r,c)
                components++
            }
        }
    }
    return components

    function helper(row, col){
        if(row < 0 || col < 0 || row >= grid.length || col >= grid[0].length || grid[row][col] == 0){
            return
        }
        grid[row][col] = 0
        helper(row - 1, col)
        helper(row + 1, col)
        helper(row, col - 1)
        helper(row, col + 1)
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows = len(grid)
        cols = len(grid[0])

        def island(r, c):
            grid[r][c] = "0"
            for dr, dc in [(1,0),(-1,0),(0,1),(0,-1)]:
                nr = r + dr
                nc = c + dc
                if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == "1":
                    island(nr, nc)

        count = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1":
                    island(r, c)
                    count += 1
    
        
        return count
